### DISCUSSION TITLE
Pass check-in `Response` to hook before returning

### DISCUSSION THREAD

**Discussion by kiancross** [pretix/pretix#3524] [Ideas]
We want to set the `require_attention` flag of some `Response`s from the check-in code based on various conditions (which probably wouldn't easily make good generic Pretix contributions).

The [`_redeem_process`](https://github.com/pretix/pretix/blob/83811c03436acedac7b3d97f4953cc3adb3d1055/src/pretix/api/views/checkin.py#L397C5-L397C20) function returns various `Response`s, and I wonder if add a hook which takes the `Response` (and some other data, e.g., the positions), it could mutate the `Response` to set the `require_attention` based on these custom conditions?

Taking this a bit further, could this hook set the `reason_explanation` to display a custom message? I assume `reason_explanation` is displayed when the `status` is `error`, but what about when `incomplete` or `ok`?

I'm aware, based on one of my previous discussions (https://github.com/pretix/pretix/discussions/3466#discussioncomment-6415551), that this might be tricky to merge, given that any changes to `Response` by a plugin wouldn't be reflected in the logic of offline check-in apps. But I guess this can be a documented limitation of the hook?

Any thoughts about adding such a hook?

**Comment by raphaelm** [comment]
I'm not very keen on adding such a hook, because it feels very dangerous for a number of reasons:

1. As you said, this would be impossible to implement this for offline check-in.
2. We try to avoid plugin hooks in code paths that tend to be very critical for performance and reliability of the system in the most critical situations.
3. Even if we were to add in such a hook (which is a big if), I believe working on the `Response` is the wrong software layer. Once we have a `Response`, it's already too late to change almost anything about it. At that point, the `Checkin` object has already been committed to the database. `require_attention` maybe the one thing we could still change, because it's not persisted, but if we add a plugin hook to change the redemption process, we'd probably want to allow changing even more than only `require_attention`. You already mentioned `reason_explanation`, which **is** persisted to the `Checkin` model, so the hook would need to run before it is saved. Possibly, the hook should even be able to change a successful checkin to an unsuccessful one or vice versa, which would require the hook to live in `perform_checkin`, although I'm not really sure *where* in that method it should live.

> Taking this a bit further, could this hook set the reason_explanation to display a custom message? I assume reason_explanation is displayed when the status is error, but what about when incomplete or ok?

reason_explanation is currently only shown on error, but we'll implement a custom text (based on product, at least) for other situations soon.

**Reply by kiancross** [reply]
Thanks for your reply. These are good points.

I think (1) and (2) can be mitigated with good documentation. Having such a hook would allow a variety of customisations to be made, which wouldn't necessarily fit well within Pretix itself. I suppose there is a tradeoff for users between having the functionality such a plugin could provide and having offline mode. But users can make this decision based on their use cases.

I agree with (3), and think the placement and signature of such a hook needs to be thought about more carefully. If you're not totally opposed, I can think more carefully about this and come back with something.

> reason_explanation is currently only shown on error, but we'll implement a custom text (based on product, at least) for other situations soon.

This is good to know.

Allowing the check-in logic to be pluggable is quite powerful and would allow lots of edge cases to be fulfilled.

**Reply by raphaelm** [reply]
> I think (1) and (2) can be mitigated with good documentation. Having such a hook would allow a variety of customisations to be made, which wouldn't necessarily fit well within Pretix itself. I suppose there is a tradeoff for users between having the functionality such a plugin could provide and having offline mode. But users can make this decision based on their use cases.

Yes, but experience shows that most users are very bad at actually reading documentation, no matter how loudly it screams at you, and not very good at understanding complex tradeoffs that involve unlikely situations. I don't think "we can put a big warning sign on it" should be a valid excuse for building inconsistent software where you need technical knowledge to predict which functionality might be suddenly missing in some cases :wink: 

> If you're not totally opposed, I can think more carefully about this and come back with something.

Yes, feel free too, as long as you emotionally prepare for the situation that it's still might be a "no" in the end :wink: 

I'd also be interested in hearing more about the real-world use case behind your request. Maybe there is a different path to the same goal that is less powerful for someone who can write their own plugins, but at the same time more powerful for anyone else :) I have not yet missed the option to hook into the checkin process during day-to-day operations.

**Reply by raphaelm** [reply]
(To give a little bit perspective on that last part: We love that pretix is customizable by folks who run their own instance and build custom logic to fulfill their special cases. But it's not our main priority. In fact, when writing down design goals for pretix back in 2014, one of the key goals was "one should not need to write code to use it for their event" :wink: Our main priority is building a general-purpose product that fulfills *most* use cases in a stable, reliable and usable way – even if we cannot support *all* use cases.)

**Reply by kiancross** [reply]
> I don't think "we can put a big warning sign on it" should be a valid excuse for building inconsistent software where you need technical knowledge to predict which functionality might be suddenly missing in some cases 😉

This is true, especially since plugin developers might not inform the user that their plugin won't work in offline mode. This would cause confusion.

One solution to this might be a toggle on the UI that the user needs select to enable the hook. This can come with a warning message explaining that plugins using such functionality won't work in offline mode.

A step beyond this might be disallowing offline mode when this toggle is enabled. In fact, perhaps plugins should be able to register a hook which disables offline mode if the functionality the plugin adds won't work in offline mode. This would remove the inconsistency between the functionality but could be a bit restrictive if users do not care that online/offline behaves slightly differently (e.g., if the plugin is only slightly changing an error message).

> Yes, feel free too, as long as you emotionally prepare for the situation that it's still might be a "no" in the end 😉

Haha, that's fine!

> I'd also be interested in hearing more about the real-world use case behind your request. Maybe there is a different path to the same goal that is less powerful for someone who can write their own plugins, but at the same time more powerful for anyone else :) I have not yet missed the option to hook into the checkin process during day-to-day operations.

Most of my posts regarding check-in have been surrounding the same use case. Mayve I should have stated precisely what this is rather perhaps proposing an [XY problem](https://xyproblem.info/)!

We have various base tickets. Some are redeemable through certain memberships and are at different costs. Others are the same price but redeemable only with access to a voucher. Essentially, we have a set of base tickets which all entitle the holder to the same thing but are obtained through different paths.

On top of this, each of these tickets has the same set of optional add-ons which can be purchased. This is things such as "VIP", "Meal", and "Entrance Drink".

We also have some hidden add-ons which staff can apply to tickets (e.g., "Accessible").

This part all works fantastically with Pretix, and configuring this is easy.

The problem comes at check-in. By default, all base-ticket holders (those without any add-ons) can use the standard entrance. Those who have purchased "VIP" can use the VIP entrance _or_ the standard entrance. Those who have purchased "Meal" or "Entrance Drink" must use the food entrance (and note that customers may have purchased _both_ add-ons). My solutions to this were https://github.com/pretix/pretix/discussions/3466 and https://github.com/pretix/pretix/discussions/3523.

Furthermore, the "Accessible" add-on provides access through any entrance, and should also display a message (e.g., 'requires attention'), so that they can be given a wristband. My solution to this was proposed in this discussion.

As you can see, this is a fairly complex setup!

> (To give a little bit perspective on that last part: We love that pretix is customizable by folks who run their own instance and build custom logic to fulfill their special cases. But it's not our main priority. In fact, when writing down design goals for pretix back in 2014, one of the key goals was "one should not need to write code to use it for their event" wink Our main priority is building a general-purpose product that fulfills most use cases in a stable, reliable and usable way – even if we cannot support all use cases.)

I completely agree with this goal: "one should not need to write code to use it for their event". But as you know, it's impossible to achieve for _everyone_:) It makes sense that your main priority is a general-purpose product that fulfils the requirements of _most_ people. But when it doesn't, it's nice to be able to write a plugin to add the missing functionality! In this regard, it reminds me a little bit of WordPress, which has been incredibly successful.

I think there's a balance to strike between making it usable for general users and allowing power-users to get what they need. A rich plugin system is probably the solution to that.

**Comment by raphaelm** [comment]
> One solution to this might be a toggle on the UI that the user needs select to enable the hook. This can come with a warning message explaining that plugins using such functionality won't work in offline mode.

This might work, but also feels inconsistent with how plugins work elsewhere.

> A step beyond this might be disallowing offline mode when this toggle is enabled. In fact, perhaps plugins should be able to register a hook which disables offline mode if the functionality the plugin adds won't work in offline mode. This would remove the inconsistency between the functionality but could be a bit restrictive if users do not care that online/offline behaves slightly differently (e.g., if the plugin is only slightly changing an error message).

Yes, I think it's too restrictive. Maybe people might prefer having the offline mode with less functionality to not having it at all. It's a risk-tradeoff that is (already, with the current capabilities) hard to understand, but still important.

> We have various base tickets. Some are redeemable through certain memberships and are at different costs. Others are the same price but redeemable only with access to a voucher. Essentially, we have a set of base tickets which all entitle the holder to the same thing but are obtained through different paths.
> 
> On top of this, each of these tickets has the same set of optional add-ons which can be purchased. This is things such as "VIP", "Meal", and "Entrance Drink".
> 
> We also have some hidden add-ons which staff can apply to tickets (e.g., "Accessible").

This **is** a fairly complex setup, and it also feels like a fairly uncommon one. There's a few things about it that probably make sense to you, but not yet to me. For example, if I have ordered a "meal and entrance drink" add-on, why can't I go through the regular entrance and get my meal and drink later? Most events would allow this.

In an ideal world (and I know we're not fully there yet), I'd structure the entire setup differently, though, from what I now know. My base products would be "Regular ticket", "Accessible ticket" and "VIP ticket". And the "different paths" (different prices, membership requirements, voucher requirements, …) should be *variations* on that ticket. In other words: The *product* should be the "what you get" and the *variation* should be the "who gets it/how you get it". 

With that, your regular/VIP/accessible entrance is solved, at least after #3521 has been merged.

I don't think add-ons will ever be a good fit for something like "VIP" or "Accessible" in most scenarios. But they are probably the correct tool for the meal and entrance drink. If you claim these at a separate place and time than check-in, this brings us back to #3523, and if meal and entrance drink are separated, can already be solved with `addon_match`. If you claim these at your first check-in, maybe even just #3526 would suffice?

**Reply by kiancross** [reply]
> This might work, but also feels inconsistent with how plugins work elsewhere.

Alternatively, plugin developers could be required to declare the use of the hook in the plugin's manifest. As a result, when enabling the plugin, a warning can be displayed. The hook will not be called for plugins which do not declare its use. This is more development effort, but it probably leads to the most consistent UX.

> Yes, I think it's too restrictive. Maybe people might prefer having the offline mode with less functionality to not having it at all. It's a risk-tradeoff that is (already, with the current capabilities) hard to understand, but still important.

Agreed.

> This is a fairly complex setup, and it also feels like a fairly uncommon one.

I thought so! This is probably why implementing some of our strange requirements as plugins makes more sense for us. It just requires Pretix to have appropriate hooks.

> There's a few things about it that probably make sense to you, but not yet to me. For example, if I have ordered a "meal and entrance drink" add-on, why can't I go through the regular entrance and get my meal and drink later? Most events would allow this.

In this case, it is because the meal is at a set time, accessible through a particular entrance on the site. Entering through the regular entrance would mean the venue for the meal would be unreachable due to crowd-control measures. Thus, it is better for us that the ticket is rejected if the customer tries to use the regular entrance so that we can direct them to the correct entrance, located a couple of minutes away. The same applies to entrance drinks.

But even if we relaxed this and allowed those with the meal/drink to enter through the regular entrance, I still think we'd end up with the problem of how to limit those _without_ the meal/drink from entering through the special entrances (because we currently have this configured as add on products).

> In an ideal world (and I know we're not fully there yet), I'd structure the entire setup differently, though, from what I now know.

Thank you very much for the rest of your reply. I need to think about this more (and probably do a test setup to properly understand it) before I can fully reply.

> With that, your regular/VIP/accessible entrance is solved, at least after https://github.com/pretix/pretix/pull/3521 has been merged.

Agreed. If we were able to move everything to product variations, #3521 solves this.

> I don't think add-ons will ever be a good fit for something like "VIP" or "Accessible" in most scenarios.

I think add-ons might be a good fit for 'accessible' in our case because it can be added to any ticket type (VIP, standard etc). It's mostly informational (which #3526 helps with), although it does provide access to a special entrance.

> But they are probably the correct tool for the meal and entrance drink. If you claim these at a separate place and time than check-in, this brings us back to https://github.com/pretix/pretix/discussions/3523

They're claimed at the same time as checkin, but through a separate entrance.

> and if meal and entrance drink are separated, can already be solved with `addon_match`

I don't think `addon_match` works, because it requires having a separate check-in list, which means the base ticket can still be redeamed at the standard entrance (and is why I also proposed #3466).

Thanks again for all your help!
